TrkDiag/src/BkgDiag_module.cc

```diff
@@ -75,121 +90,115 @@ namespace mu2e
       const BkgClusterHitCollection *_bkghitcol;
 
       // background diagnostics
-      TTree* _bdiag,*_bdiag2;
-      Int_t _iev;
-      CLHEP::Hep3Vector _cpos;
-      Float_t _ctime;
-      Float_t _mindt, _mindrho;
-      Bool_t _isbkg, _isref, _isolated, _stereo;
-      Int_t _cluIdx, _nactive, _nchits, _nshits, _nstereo, _nsactive, _nbkg;
+      TTree* _bcdiag,*_bhdiag;
+      int _iev = 0;
+      XYZVectorF _cpos;
+      float _ctime = 0;
+      float _mindt = 0;
+      float _mindrho = 0;
+      bool _isbkg = false;
+      bool _isref = false;
+      bool _isolated = false;
+      bool _stereo = false;
+      int _cluIdx, _nactive, _nchits, _nshits, _nstereo, _nsactive, _nbkg;
       // BkgQual vars
       float _bkgqualvars[BkgQual::n_vars];
-      Int_t _mvastat;
-      Float_t _mvaout;
+      int _mvastat;
+      float _mvaout;
 
       // MC truth variables
-      Int_t _ppid, _ppdg, _pgen, _pproc, _ncontrib, _icontrib[512];
-      Float_t _pmom;
-      Int_t _nconv, _ndelta, _ncompt, _ngconv, _nebkg, _nprot, _nprimary;
+      int _ppid, _ppdg, _pgen, _pproc, _ncontrib, _icontrib[512];
+      float _pmom;
+      int _nconv, _ndelta, _ncompt, _ngconv, _nebkg, _nprot, _nprimary;
       std::vector<BkgHitInfo> _bkghinfo;
 
-      Int_t   _nindex,_hitidx[8192];
```

Comment:
I don't believe clusters with a single hits are saved, so we would need this tree to store the information of *all* hits. I saved some information, but the alternative would be to make a tree with a single BkhHitInfo branch, and record the info of all hits. Then in your original tree, you just have a pointer to these hits. Or we keep it as is, or we make dummy clusters for unassociated hits. Pick you poison

Reply:
There are 2 trees, one with cluster information (and information about the hits in that cluster), and one generic tree for all hits.  I'm not sure which you mean by 'original tree'.  Trees can't save pointers, so I also don't know what you mean by that.  Are you worried about the length of the array?  Root overwrites that in the streamer.  If it's something else please clarify.  This is an old section of the code and, though it's ugly, as far as I can tell it does the right thing.

Reply:
Sorry for the confusion. The BkgDiag tree has info about clusters and the hits that make them (note that this is only a subset of all hits). The BkgDiag2 tree has summary info about all the hits in the event. One could only store the hit information in the BkgDiag2 tree, and save the indices of the hits forming each cluster in BkgDiag. This would save space, but introduce coupling between the two trees, and this will probably cause problems with other scripts now that I think about it. So we will leave it this way.